elbepack/cdroms.py

```diff
@@ -56,6 +57,12 @@ def mk_source_cdrom(components, codename,
     for component in components.keys():
         rfs, cache, pkg_lst = components[component]
         logging.info('Adding %s component', component)
+
+        forbiddenSrcPackages = []
+        for name, _ in cache.get_corresponding_source_packages(forbiddenPackages):
```

Comment:
This doesn't seem correct.

If multiple packages share the same `Build-Using` and only one of the binary packages is excluded from the source cdrom, then that exclusion will override all others.
This is especially problematic if some base Debian packages have been used to build some excluded package.

The correct solution seems to need to go away from "on_src_cd" to some external explicit list.
However as a good-enough solution it should be fine to add a (keyword only) parameter to `get_corresponding_source_packages()` to inhibit `Build-Using` handling.

Reply:
Ok I didn't had in mind, that the build dependencies are then also excluded.

And I agree a separate list would make the things more visible.

I'm not 100% sure if I understand the proposed good-enough solution correctly.

You mean something like the following approach?

```python
forbidden_src_packages = set()
for name, _ in cache.get_corresponding_source_packages(forbidden_packages, parse_build_depends=False):
    forbidden_src_packages.add(name)
```

```python
def get_corresponding_source_packages(cache, pkg_lst=None, parse_build_depends=True):

    if pkg_lst is None:
        pkg_lst = {p.name for p in cache if p.is_installed}

    src_set = set()

    for pkg in pkg_lst:
        if isinstance(pkg, str):
            version = cache[pkg].installed or cache[pkg.candidate]
        elif isinstance(pkg, PackageBase):
            version = cache[pkg.name].versions[pkg.installed_version]

        src_set.add((version.source_name, version.source_version))

        if parse_build_depends:
            for name, ver in parse_built_using(version.record.get('Built-Using')):
                src_set.add((name, ver))

    return list(src_set)
```

```python
def get_corresponding_source_packages(self, pkg_lst=None, parse_build_depends=True):
    return get_corresponding_source_packages(self.cache, pkg_lst, parse_build_depends)
```